test/runnable/test17924.d

```diff
@@ -0,0 +1,25 @@
+// PERMUTE_ARGS:
+
+void main()
+{
+    int x = 0;
+    int y = 1;
+    assert((x ?: y) == y);
+    assert((y ?: x) == y);
+
+    int* p;
+    int* u;
+    assert((p ?: u) == u);
+    assert((u ?: p) == u);
+
+    class C
+    {
+        int a;
+        this(int a) { this.a = a; }
+    }
+
+    C inst = new C(1);
+    C not_inst;
+    assert((not_inst ?: inst).a == inst.a);
+    assert((inst ?: not_inst).a == inst.a);
+}
```

Comment:
To we need to add tests for other types that implicitly converts to `bool`? Like function pointers, delegates, structs with `opCast`.

Reply:
I don't really think that it is necessary that case is already checked for the ternary operator. This PR just modifies the parser, so everything else is treated exactly as in the ternary operator case.